### PR TITLE
Fix bug in add_template with existing link of same name.

### PR DESCRIPTION
**Pull request by andrewmwells-amazon**
## Description of changes
Fix bug in add_template with existing link of same name.

## Issue #, if available
Related to https://github.com/cedar-policy/cedar/issues/455

## Checklist for requesting a review

The change in this PR is (choose one, and delete the other options):

- [ ] A breaking change requiring a major version bump to `cedar-policy` (e.g., changes to the signature of an existing API).
- [ ] A backwards-compatible change requiring a minor version bump to `cedar-policy` (e.g., addition of a new API).
- [x] A bug fix or other functionality change requiring a patch to `cedar-policy`.
- [ ] A change "invisible" to users (e.g., documentation, changes to "internal" crates like `cedar-policy-core`, `cedar-validator`, etc.)
- [ ] A change (breaking or otherwise) that only impacts unreleased or experimental code.

I confirm that this PR (choose one, and delete the other options):

- [x] Updates the "Unreleased" section of the CHANGELOG with a description of my change (required for major/minor version bumps).
- [ ] Does not update the CHANGELOG because my change does not significantly impact released code.

I confirm that [`cedar-spec`](https://github.com/cedar-policy/cedar-spec) (choose one, and delete the other options):

- [x] Does not require updates because my change does not impact the Cedar Dafny model or DRT infrastructure.
- [ ] Requires updates, and I have made / will make these updates myself. (Please include in your description a timeline or link to the relevant PR in `cedar-spec`, and how you have tested that your updates are correct.)
- [ ] Requires updates, but I do not plan to make them in the near future. (Make sure that your changes are hidden behind a feature flag to mark them as experimental.)
- [ ] I'm not sure how my change impacts `cedar-spec`. (Post your PR anyways, and we'll discuss in the comments.)

## Disclaimer

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: cedar-policy-core/src/ast/policy_set.rs [hunk 1]
```diff
@@ -161,7 +161,11 @@ impl PolicySet {
     }
 
     /// Add a template to the policy set.
+    /// If a link with the same name already exists, this will error.
```
Comment (cdisselkoen) [inline cedar-policy-core/src/ast/policy_set.rs:164]: what about if a static policy with the same name already exists?  or if a template with the same name already exists?  I'm sure these are both errors, but we can/should call that out in the doc comment

File: cedar-policy/src/api.rs [hunk 2]
```diff
@@ -3662,6 +3662,36 @@ mod policy_set_tests {
...
+            Some("policy3".into()),
+            "permit(principal == ?principal, action, resource);",
+        )
+        .expect("Template Parse Failure");
+
+        assert_matches!(
+            pset.add_template(template),
+            Err(PolicySetError::AlreadyDefined)
+        );
+        //Should not panic
```
Comment (cdisselkoen) [inline cedar-policy/src/api.rs:3693]: do we already have tests that add a template with the same name as an existing static policy, or as an existing template, throws this error?

**Review by cdisselkoen — APPROVED** [review]

File: cedar-policy-core/src/ast/policy_set.rs (hunk 1 above)
Comment (andrewmwells-amazon) [reply cedar-policy-core/src/ast/policy_set.rs:164]: Fair point.

File: cedar-policy/src/api.rs (hunk 2 above)
Comment (andrewmwells-amazon) [reply cedar-policy/src/api.rs:3693]: I didn't add tests for it but from reading the code I'm fairly confident (famous last words)

Comment (cdisselkoen) [reply cedar-policy/src/api.rs:3693]: Adding tests for that is probably reasonable just to ensure we don't break it in the future

**Comment by cdisselkoen** [comment]
Oh another comment -- are we confident this change isn't breaking?  in the sense that, does Cedar 2.4.2 allow a template and a link to have the same name, and now we're disallowing (breaking) that?

**Comment by andrewmwells-amazon** [comment]
> Oh another comment -- are we confident this change isn't breaking? in the sense that, does Cedar 2.4.2 allow a template and a link to have the same name, and now we're disallowing (breaking) that?

Definitely breaking but I think the old behavior is buggy. I don't see why `link; add_template` would be allowed but `add_template; link` would be an error. The order in which you add policies shouldn't change behavior.

**Comment by andrewmwells-amazon** [comment]
Since the 3.0 release is coming soon, I merged into `main` instead (https://github.com/cedar-policy/cedar/pull/456)

**Pull request closed without merging**